zipkin-cassandra-core/src/main/java/org/twitter/zipkin/storage/cassandra/Repository.java

```diff
@@ -707,6 +744,131 @@ private String debugInsertTraceIdByAnnotation(ByteBuffer annotationKey, long tim
                 .replace(":ttl_", String.valueOf(ttl));
     }
 
+    private class DurationRow {
+        Long trace_id;
+        Long duration;
+        Long timestamp;
+        DurationRow(Row row) {
+            trace_id = row.getLong("tid");
+            duration = row.getLong("d");
+            timestamp = row.getLong("ts");
+        }
+        public String toString() {
+            return String.format("trace_id=%d, duration=%d, timestamp=%d", trace_id, duration, timestamp);
+        }
+    }
+
+    /**
+     * @return Returns a map of trace id -> timestamp
+     */
+    public ListenableFuture<Map<Long, Long>> getTraceIdsByDuration(String serviceName, String spanName,
+                                                                   long minDuration, long maxDuration,
+                                                                   long endTs, long startTs, int limit) {
+        int startBucket = durationIndexBucket(startTs);
+        int endBucket = durationIndexBucket(endTs);
+        try {
+            if (startBucket > endBucket) {
+                throw new IllegalArgumentException("Start bucket (" + startBucket + ") > end bucket (" + endBucket + ")");
+            }
+            IntFunction<ListenableFuture<List<DurationRow>>> oneBucketQuery = (bucket) -> {
+                BoundStatement bound = selectTraceIdsBySpanDuration.bind()
+                        .setString("service_name", serviceName)
+                        .setString("span_name", spanName == null ? "" : spanName)
```

Comment:
Can't do much about that, if the correct span name is not available/incorrect at the time of the insert. It will only get indexed by the service name.